### PR TITLE
[DNM]add entry cache

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
by busyjay.

### CODE REVIEW

File: src/util/mod.rs [hunk 1]
```diff
@@ -433,6 +438,7 @@ pub fn monitor_threads<S: Into<String>>(_: S) -> io::Result<()> {
 }
 
 /// A simple ring queue with fixed capacity.
+#[derive(Clone, Debug)]
```
Comment (siddontang) [inline src/util/mod.rs]: please use another PR for RingQueue

File: src/util/mod.rs [hunk 2]
```diff
@@ -49,20 +47,27 @@ mod thread_metrics;
...
-pub fn limit_size<T: Message + Clone>(entries: &mut Vec<T>, max: u64) {
-    if entries.is_empty() {
-        return;
-    }
-
-    let mut size = Message::compute_size(&entries[0]) as u64;
-    let mut limit = 1usize;
-    while limit < entries.len() {
-        size += Message::compute_size(&entries[limit]) as u64;
+pub fn get_limit_at_size<'a, T: Message + Clone, I: IntoIterator<Item = &'a T>>(entries: I,
```
Comment (siddontang) [inline src/util/mod.rs]: any test to cover all conditions?

File: src/raftstore/store/peer_storage.rs [hunk 3]
```diff
@@ -103,6 +104,8 @@ pub fn first_index(state: &RaftApplyState) -> u64 {
     state.get_truncated_state().get_index() + 1
 }
 
+const DEFAULT_CACHE_SIZE: usize = 3000;
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: If we have many regions like > 10000 in one store, and in most cases, there are only some active regions, so use 3000 for all regions seem wasteful. 

Maybe we can create the RingQueue with a small capacity and a large max size.

File: src/raftstore/store/peer_storage.rs [hunk 4]
```diff
@@ -311,10 +317,27 @@ impl PeerStorage {
 
     pub fn entries(&self, low: u64, high: u64, max_size: u64) -> raft::Result<Vec<Entry>> {
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: Do we have enough tests to cover all the conditions?

File: src/util/mod.rs [hunk 5]
```diff
@@ -49,20 +48,29 @@ pub mod collections;
...
-    let mut size = Message::compute_size(&entries[0]) as u64;
-    let mut limit = 1usize;
-    while limit < entries.len() {
-        size += Message::compute_size(&entries[limit]) as u64;
+pub use fnv::{FnvHashMap as HashMap, FnvHashSet as HashSet, FnvBuildHasher as BuildHasherDefault};
+
+pub fn get_limit_at_size<'a, T: Message + Clone, I: IntoIterator<Item = &'a T>>(entries: I,
+                                                                                max: u64)
+                                                                                -> usize {
+    let mut iter = entries.into_iter();
```
Comment (siddontang) [inline src/util/mod.rs]: if max is NO_LIMIT, I think we can return the limit directly, no need to compute the sizes. 

And in our raft, most of the places use NO_LIMIT.

File: src/util/mod.rs [hunk 6]
```diff
@@ -523,6 +595,57 @@ mod tests {
...
+                queue.push(num);
+            }
+            let exp: Vec<_> = (0..cap + pos).collect();
+            // now queue's start index is at pos.
+            let v = queue.to_vec(0..queue.len());
+            assert_eq!(*v, exp[pos..]);
+            for i in 0..cap {
+                queue.drain(cap - i - 1..cap - i);
+                let res = queue.to_vec(0..queue.len());
+                assert_eq!(*res, exp[pos..pos + cap - i - 1]);
```
Comment (hhkbp2) [inline src/util/mod.rs]: I thought it would be easier to understand to initialize `exp` to be `(pos.. pos +cap)` and to use it with start index set to 0.

**Comment by ngaut** [comment]
Let's keep this PR with [Do not merge] state. Since we will merge multiple operation into a single raft log entry, maybe there is no need to keep raft log entry cache.

**Pull request closed without merging**